File: stateReader.ts
import * as fs from 'fs';

interface TerraformState {
    version: number;
    terraform_version: string;
    resources: Resource[];
}

interface Resource {
    name: string;
    mode: ResourceMode;
}

enum ResourceMode {
    managed = 'managed',
    other = 'other', // Add other resource modes as needed
}

class StateFileReader {
    private file_path: string;
    private data: TerraformState;

    constructor(file_path: string) {
        this.file_path = file_path;
        this.data = this.load_file();
    }

    private load_file(): TerraformState {
        try {
            const fileContent = fs.readFileSync(this.file_path, 'utf8');
            const data = JSON.parse(fileContent);
            return data as TerraformState;
        } catch (error) {
            console.error(`Error loading or validating state file: ${error}`);
            return {
                version: 0,
                terraform_version: '',
                resources: [],
            };
        }
    }

    list_resources(): string[] {
        return this.data.resources.map(resource => resource.name);
    }

    get_managed_resources(): Resource[] {
        return this.data.resources.filter(resource => resource.mode === ResourceMode.managed);
    }

    resource_exists(resource_name: string): boolean {
        return this.data.resources.some(resource => resource.name === resource_name);
    }

    get_resource_added_date(resource_name: string): string | null {
        for (const resource of this.data.resources) {
            if (resource.name === resource_name) {
                // TODO: Add functionality to get resource added date if possible
                return resource.model_dump(); // Replace with actual implementation
            }
        }
        return null;
    }
}

File: state_reader.py
import json
import typing

import models
import pydantic


class StateFileReader:
    def __init__(
        self,
        file_path: str,
    ) -> None:
        """
        Initializes the StateFileReader instance.

        Args:
            file_path (str): The path to the state file.
        """
        self.file_path = file_path
        self.data = self.load_file()
    
    def load_file(
        self,
    ) -> models.TerraformState:
        """
        Loads the state file and returns its content as a TerraformState object.

        Returns:
            models.TerraformState: The loaded and validated state data.
        """
        try:
            with open(self.file_path, 'r') as file:
                data = json.load(file)
    
            return models.TerraformState(**data)
        
        except (
            FileNotFoundError, 
            json.JSONDecodeError, 
            pydantic.ValidationError
        ) as e:
            print(
                f'Error loading or validating state file: {e}',
            )
            
            return models.TerraformState(
                version=0,
                terraform_version='',
                resources=[],
            )
    
    def list_resources(
        self,
    ) -> typing.List:
        """
        Lists the names of resources in the loaded state file.

        Returns:
            List[str]: A list of resource names.
        """
        return [
            resource.name 
            for resource in self.data.resources
        ]
        
    def get_managed_resources(
        self,
    ):
        return [
            resource for resource in self.data.resources
            if resource.mode == models.ResourceMode.managed
        ]
    
    def resource_exists(
        self,
        resource_name: str,
    ) -> bool:
        """
        Checks if a resource exists in the state file.

        Args:
            resource_name (str): The name of the resource to check.

        Returns:
            bool: True if the resource exists, False otherwise.
        """
        return any(
            resource.name == resource_name 
            for resource in self.data.resources
        )
    
    def get_resource_added_date(
        self,
        resource_name: str,
    ) -> typing.Optional[str]:
        """
        Retrieves the added date of a specific resource, if available.

        Args:
            resource_name (str): The name of the resource.

        Returns:
            typing.Optional[str]: The date the resource was added, or None if not available.
        """
        for resource in self.data.resources:
            if resource.name == resource_name:
                # TODO: add functionality to get resource added date if possible
                return resource.model_dump()
    
        return None
